_studio/shared/umc/codec/mpeg2_dec/src/umc_mpeg2_decoder_va.cpp

```diff
@@ -0,0 +1,156 @@
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+#include "umc_defs.h"
+#if defined (MFX_ENABLE_MPEG2_VIDEO_DECODE)
+
+#include <algorithm>
+
+#include "umc_va_base.h"
+#include "umc_mpeg2_decoder_va.h"
+#include "umc_mpeg2_va_packer.h"
+
+namespace UMC_MPEG2_DECODER
+{
+    MPEG2DecoderVA::MPEG2DecoderVA()
+        : m_va(nullptr)
+    {}
+
+    UMC::Status MPEG2DecoderVA::SetParams(UMC::BaseCodecParams* info)
+    {
+        if (!info)
+            return UMC::UMC_ERR_NULL_PTR;
+
+        MPEG2DecoderParams* dp =
+            DynamicCast<MPEG2DecoderParams, UMC::BaseCodecParams>(info);
+        if (!dp)
+            return UMC::UMC_ERR_INVALID_PARAMS;
+
+        if (!dp->pVideoAccelerator)
+            return UMC::UMC_ERR_NULL_PTR;
+
+        m_va = dp->pVideoAccelerator;
+        m_packer.reset(Packer::CreatePacker(m_va));
+
+        SetDPBSize(m_params.async_depth + NUM_REF_FRAMES);
+
+        return UMC::UMC_OK;
+    }
+
+    UMC::Status MPEG2DecoderVA::Submit(MPEG2DecoderFrame& frame, uint8_t fieldIndex)
+    {
+        VM_ASSERT(m_va);
+        UMC::Status sts = UMC::UMC_OK;
+
+        sts = m_va->BeginFrame(frame.GetMemID());
+        if (sts != UMC::UMC_OK)
+            return sts;
+
+        m_packer->PackAU(frame, fieldIndex);
+
+        sts = m_va->EndFrame();
+        if (sts != UMC::UMC_OK)
+            return sts;
+
+        return sts;
+    }
+
+    inline bool InProgress(MPEG2DecoderFrame const& frame)
+    {
+        return frame.DecodingStarted() && !frame.DecodingCompleted();
+    }
+
+    void MPEG2DecoderVA::AllocateFrameData(UMC::VideoDataInfo const& info, UMC::FrameMemID id, MPEG2DecoderFrame& frame)
+    {
+        VM_ASSERT(id != UMC::FRAME_MID_INVALID);
+
+        UMC::FrameData fd;
+        fd.Init(&info, id, m_allocator);
+
+        frame.Allocate(&fd);
+
+        UMC::FrameData* fd2 = frame.GetFrameData();
+        VM_ASSERT(fd2);
+        fd2->m_locked = true;
+    }
+
+    bool MPEG2DecoderVA::QueryFrames(MPEG2DecoderFrame& frame)
+    {
+        DPBType decode_queue;
+
+        {
+            std::unique_lock<std::mutex> l(m_guard);
+
+            // form frame queue in decoded order
+            std::for_each(m_dpb.begin(), m_dpb.end(),
+                [&decode_queue](MPEG2DecoderFrame * f)
+                {
+                    if (InProgress(*f))
+                        decode_queue.push_back(f);
+                }
+            );
+        }
+
+        decode_queue.sort([](MPEG2DecoderFrame const* f1, MPEG2DecoderFrame const* f2) { return f1->decOrder < f2->decOrder; });
+
+        decode_queue.remove_if(
```

Comment:
May be don't add earlier frames to decode queue from the beginning?

Reply:
ok